drivers: usb: dc: stm32: Added implementation to functions

JonathanVanWin opened the pull request:
Added the attached state to the USB stm32 driver,
and logging for when the device is not attached.
Also added implementation for the **usb_dc_detach**,
**usb_dc_reset** and **usb_dc_ep_flush** functions.
The **usb_dc_ep_flush** uses HAL_PCD_EP_Flush under the hood, that
uses the ll_usb. It doesn't return an error if a timeout occurred (T/RXFFLSH bit stays 1),
and it also doesn't check before doing flush if the core is reading/writing to the FIFOs.
The usb_dc_ep_disable should also AFAIK flush the FIFO, but the HAL_PCD_EP_Close
doesn't do it. I think that the usb_dc_ep_set_stall should also flush the TxFIFO (for IN endpoints).
The **usb_dc_reset** will return timeout, after **USB_STM32_CORE_RST_TIMEOUT_US** microseconds,
that parameter may be converted to a Kconfig configuration, or changed as needed.
Also the reference manual says we need to wait 3 PHY clock cycles, that I don't know how much that is,
and how to calculate, so I just took what I saw in another driver. Would like to be corrected here.
I ran a simple project that enables and disables the USB device, and resets it.
Checked with lsusb, and dmesg, and for enable/disable, show the events in dmesg,
but for reset didn't saw anything.
Did not check flush, but it is the same implementation of flush as in the ll drivers, just without checking for
timeout.
Are we supposed to see an event when we reset the USB device? It didn't return a timeout error, so I believe it works
I based my implementation on the reference manual, HAL's and ll's, and other drivers, but I believe there may still be things missing, so would happy to get some critique.

Closes #36211

Signed-off-by: Prema Jonathan van Win <jonathanvanwin@gmail.com>

JonathanVanWin commented:
Related issue: #36211

jfischer-no commented:
From the commit message:
>Added the attached state to the usb stm32 driver,
and logging for when the device is not attached.
Also added implementation for the usb_dc_detach,
usb_dc_reset and usb_dc_ep_flush functions.

If it seems that there are several independent changes, like the one with usb_dc_reset and direct controller register access, can you split it into separate commits, please?

JonathanVanWin committed:
drivers: usb: dc: stm32: Added attached state

Added the attached state to the driver state,
and check before doing something with the endpoints,
if the usb device is attached or not.

Signed-off-by: Prema Jonathan van Win <jonathanvanwin@gmail.com>

JonathanVanWin committed:
drivers: usb: dc: stm32: Added log messages

If the endpoint state does not exist (invalid
endpoint), or the usb device is detached, and
we are trying to do something with the endpoint,
show an error message.

Signed-off-by: Prema Jonathan van Win <jonathanvanwin@gmail.com>

JonathanVanWin committed:
drivers: usb: dc: stm32: Added usb_dc_detach

Added the implementation for usb_dc_detach.
Now the attached state has a purpose, because before
there was no way to detach the usb device.

Signed-off-by: Prema Jonathan van Win <jonathanvanwin@gmail.com>

JonathanVanWin committed:
drivers: usb: dc: stm32: Added usb_dc_ep_flush

Added the implementation for usb_dc_ep_flush.
The HAL_PCD_EP_FLush function used in this driver,
uses under the hood the stm32xxxx_ll_usb driver, that
returns HAL_TIMEOUT if after a time period the TXFFLSH is
still set. The HAL_PCD doesn't return the timeout error, but
always HAL_OK.
The ll driver also doesn't check before doing flush if the
core is reading/writing to the FIFOs (it should check the NAK
effective and AHBIDL bits before the flush).

Signed-off-by: Prema Jonathan van Win <jonathanvanwin@gmail.com>

JonathanVanWin committed:
drivers: usb: dc: stm32: Added usb_dc_reset

Added the implementation for usb_dc_reset.
It waits in a loop 1 microsecond till timeout, or
till (in the first loop) the AHBIDL bit is set or till
(in the second loop) the core soft reset bit is set.
Maybe we should create a config for the max timeout.
According to the reference manual we should wait 3 phy clocks,
that I don't know how much that is, and how to calculate it,
so I just set it to 100 microseconds.

Signed-off-by: Prema Jonathan van Win <jonathanvanwin@gmail.com>

JonathanVanWin commented:
> can you split it into separate commits, please?
Done, split it into five different commits
- The attached state
- The logging
- And three more commits, each one for the function I added the implementation to.

jameswalmsley commented:
@JonathanVanWin I've pulled in your PR and have tested the usb_dc_ep_flush() functionality.

I've added a line:

```
k_sem_give(&ep_state->write_sem);
```

To my code to allow another write to the USB EP.

I imagine thats not how it should be done. When and from which context should usb_dc_ep_flush() be called.

For context I want to cancel a transfer to the host if a timeout has been exceeded, and need to flush the endpoint so new transfers can take place.

erwango commented:
@loicpoulain Would you mind review this one ?

loicpoulain commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -144,6 +144,7 @@ struct usb_dc_stm32_state {
 	struct usb_dc_stm32_ep_state out_ep_state[USB_NUM_BIDIR_ENDPOINTS];
 	struct usb_dc_stm32_ep_state in_ep_state[USB_NUM_BIDIR_ENDPOINTS];
 	uint8_t ep_buf[USB_NUM_BIDIR_ENDPOINTS][EP_MPS];
+	uint8_t attached;
```
bool?

JonathanVanWin replied:
Will change

loicpoulain commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -444,6 +445,9 @@ static int usb_dc_stm32_init(void)
 #endif /* USB */
 
+	/* Disable soft disconnect */
+	USB_DEVICE_REG->DCTL &= ~USB_OTG_DCTL_SDIS;
```
Isn't it something do-able via the HAL? I'm a bit reluctant writing dev register directly here, is it compatible with all STM USB controllers?

JonathanVanWin replied:
Okay previous comment was for the detach, but for the attach we want:
```c
HAL_StatusTypeDef HAL_PCD_DevConnect(PCD_HandleTypeDef *hpcd)
{
  USB_OTG_GlobalTypeDef *USBx = hpcd->Instance;

  __HAL_LOCK(hpcd);

  if ((hpcd->Init.battery_charging_enable == 1U) &&
      (hpcd->Init.phy_itface != USB_OTG_ULPI_PHY))
  {
    /* Enable USB Transceiver */
    USBx->GCCFG |= USB_OTG_GCCFG_PWRDWN;
  }
  (void)USB_DevConnect(hpcd->Instance);
  __HAL_UNLOCK(hpcd);

  return HAL_OK;
}

/* or the lower one */
HAL_StatusTypeDef  USB_DevConnect(USB_OTG_GlobalTypeDef *USBx)
{
  uint32_t USBx_BASE = (uint32_t)USBx;

  /* In case phy is stopped, ensure to ungate and restore the phy CLK */
  USBx_PCGCCTL &= ~(USB_OTG_PCGCCTL_STOPCLK | USB_OTG_PCGCCTL_GATECLK);

  USBx_DEVICE->DCTL &= ~USB_OTG_DCTL_SDIS;

  return HAL_OK;
}
```
I'm going to assume you agree, and push the changes, if someone disagrees, or sees that it may have consequences can always comment.

loicpoulain commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -959,11 +963,15 @@ int usb_dc_ep_mps(const uint8_t ep)
+
+	/* Enable soft disconnect */
+	USB_DEVICE_REG->DCTL |= USB_OTG_DCTL_SDIS;
+
```
Why not using USB_DevDisconnect() ? which seems HAL equivalent

JonathanVanWin replied:
See previous comment.

loicpoulain commented:
@JonathanVanWin, The general idea is fine, but the driver normally relies on STM32 HAL so that USB controller variants are mostly abstracted, here you bring some direct controller register accesses which may not be compatible with all variants, and possibly cause troubles. For such code, use the HAL when possible (e.g. HAL_PCD_DevDisconnect) or flag it.

loicpoulain commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -50,6 +50,7 @@ LOG_MODULE_REGISTER(usb_dc_stm32);
 #endif
 
 #define USB_BASE_ADDRESS	DT_INST_REG_ADDR(0)
+#define USB_DEVICE_REG	((USB_OTG_DeviceTypeDef *)(USB_BASE_ADDRESS + USB_OTG_DEVICE_BASE))
```
This seems instance specific.

JonathanVanWin replied:
Okay, removed the USB_DEVICE_REG, but unfortunately the USB_CORE_REG will have to stay for now, because there is no implementation I could find. Only the USB_CoreReset, but it is static, so not exported.

JonathanVanWin commented:
@loicpoulain Okay, understood.
Will fix the direct register access.
I'm not currently next to a computer, so expect to see the CR fixed by Friday.

loicpoulain commented:
Hi @JonathanVanWin , gentle ping.

erwango commented:
@JonathanVanWin Do you think you'll be able to complete this PR ?

JonathanVanWin commented:
Yes, I will fix it ASAP.

JonathanVanWin commented:
Should I also add what @[jameswalmsley](https://github.com/jameswalmsley) mentioned (https://github.com/zephyrproject-rtos/zephyr/pull/36848#issuecomment-914399057)

holgerschurig commented:
Your patch assumes OTG hardware. But on my STM32F303, there is none. Therefore this line

```
while (!(USB_CORE_REG->GRSTCTL & USB_OTG_GRSTCTL_AHBIDL)) {
```

and similar ones won't work. I'm afraid you need to abstract the hardware even more ...

erwango commented:
@JonathanVanWin Would you be able to rebase before @loicpoulain can review ?

JonathanVanWin commented:
> @JonathanVanWin Would you be able to rebase before @loicpoulain can review ?

I will try to do it this weekend.

JonathanVanWin commented:
> > @JonathanVanWin Would you be able to rebase before @loicpoulain can review ?
> 
> I will try to do it this weekend.

@erwango Done

erwango commented:
@loicpoulain would you have time to review this once again ?

erwango commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -1044,7 +1045,36 @@ int usb_dc_detach(void)
+	uint32_t cnt = 0U;
+
+	/* Wait for AHB master idle state. */
+	while (!(USB_CORE_REG->GRSTCTL & USB_OTG_GRSTCTL_AHBIDL)) {
```
This is only valid on series that have a USB OTG (`st,stm32-otgfs`), which excludes F0/F3/G0/G4/L0/(some)L4.
So either it's possible to abstract for other kind of USB HW blocks, either you need to restrict this code under this compatible for now.

erwango reviewed (changes requested):
One comment on my side.
Otherwise LGTM

Light411 commented:
I used most of this PR to patch my usb_dc_stm32.c file. I add to remove the part highlighted by @erwango because it is not implemented on G4 family. 
Maybe we could keep the implementation of usb_dc_reset for later as it seems more complicated than expected ?

erwango commented:
Fixed by https://github.com/zephyrproject-rtos/zephyr/pull/51473
